src/main/groovy/com/google/gcloud/GCloudPlugin.groovy

```diff
@@ -1,33 +1,131 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package com.google.gcloud
-
-import com.google.appengine.AppEnginePlugin
-import com.google.appengine.AppEnginePluginExtension
-import com.google.gcloud.task.GCloudAppDeployTask
-import com.google.gcloud.task.GCloudAppRunTask
+import com.google.gcloud.dsl.GCloudConfigurable
+import com.google.gcloud.dsl.GCloudPluginExtension
+import com.google.gcloud.task.app.DeployTask
+import com.google.gcloud.task.app.RunTask
+import com.google.gcloud.task.app.StopTask
+import org.gradle.api.Action
 import org.gradle.api.Incubating
 import org.gradle.api.Plugin
-import org.gradle.api.Project;
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
 
+import javax.inject.Inject
 /**
  * <p>A {@link Plugin} that provides tasks for generic use of gcloud in gradle and
  * for uploading and running of Google App Engine projects via gcloud.</p>
  *
- * @author Appu Goundan
  */
 @Incubating
 public class GCloudPlugin implements Plugin<Project> {
 
-    static final String TASK_GCLOUD_APP_RUN = "gcloudAppRun"
-    static final String TASK_GCLOUD_APP_DEPLOY = "gcloudAppDeploy"
+    static final String TASK_GCLOUD_RUN = "gcloudRun"
+    static final String TASK_GCLOUD_STOP = "gcloudStop"
+    static final String TASK_GCLOUD_DEPLOY = "gcloudDeploy"
+    static final String TASK_GROUP_GCLOUD = "Google gcloud experimental"
+
+    final ToolingModelBuilderRegistry registry;
+    final Instantiator instantiator;
+    Project project;
+    GCloudPluginExtension extension;
+
+    @Inject
+    public GCloudPlugin(Instantiator instantiator, ToolingModelBuilderRegistry registry) {
+        this.instantiator = instantiator
+        this.registry = registry
+    }
 
     @Override
-    void apply(Project project) {
+    void apply(Project p) {
+        project = p;
+        extension = project.extensions.create('gcloud', GCloudPluginExtension, project, instantiator)
+
+        createDynamicTasks()
+    }
 
-        project.apply plugin: "appengine"
+    void createDynamicTasks() {
+        project.afterEvaluate() {
+            extension.configurations.each { GCloudConfigurable configuration ->
 
-        project.extensions.create('gcloud', GCloudPluginExtension, project)
+                // deploy task
+                String deployTaskName = TASK_GCLOUD_DEPLOY + configuration.name.capitalize();
```

Comment:
Does it make sense to add a spacer between the gcloud* and the config so when it is easier to read or type?